src/conversation_engine/engine.py

```diff
@@ -0,0 +1,409 @@
+        self._robot_name = robot_name
+
+        self._user_to_robot_sub = rospy.Subscriber("user_to_robot", String, self._handle_user_to_robot)
+        self._robot_to_user_pub = rospy.Publisher("robot_to_user", String, queue_size=10)
+
+        self._latest_feedback = None
+
+        rospy.logdebug("Started conversation engine")
+
+    def __del__(self):
+        self._stop()
+
+    def _handle_user_to_robot(self, msg):
+        rospy.loginfo("_handle_user_to_robot('{}')".format(msg))
+
+        text = sanitize_text(msg.data)
+
+        if self._handle_special_commands(text):
+            return
+
+        if self._state.state == ConversationState.IDLE:
+            # parse command and send_goal
+            self._handle_command(text)
+        elif self._state.state == ConversationState.WAIT_FOR_USER:
+            # Update semantics and send updated goal
+            self._handle_additional_info(text)
+        elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+            # User must wait for robot/action server to reply, cannot handle user input now
+            self._handle_user_while_waiting_for_robot(text)
+        elif self._state.state == ConversationState.ABORTING:
+            self._handle_user_while_aborting(text)
+
+    def _handle_special_commands(self, text):
+        """Check for special commands that should not be parsed further
+        @returns bool indicating if the text is a special command"""
+
+        stop_words = ["stop", "cancel", "quit", "reset"]
+        if any([word for word in stop_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'):".format(text))
+            if self._state.state == ConversationState.IDLE:
+                self._robot_to_user_pub.publish(random.choice(["I'm not busy",
+                                                               "I'm not doing anything"]))
+                self._say_ready_for_command()
+            elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+                self._stop()
+            elif self._state.state == ConversationState.WAIT_FOR_USER:
+                self._robot_to_user_pub.publish(random.choice(["I'm waiting for you, there's nothing to stop",
+                                                               "I can't stop, you stop!"]))
+
+                self._start_new_conversation()
+            elif self._state.state == ConversationState.ABORTING:
+                self._robot_to_user_pub.publish(random.choice(["I'm already stopping, gimme some time!"]))
+            return True
+
+        kill_words = ["sudo kill"]
+        if any([word for word in kill_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'): kill action server".format(text))
+
+            self._action_client.cancel_all_async()
+
+            self._robot_to_user_pub.publish(random.choice(["Woah, sorry dude!"]))
+            os.system("rosnode kill /state_machine")
```

Comment:
Isn't there a proper rospy API to handle this? This is a bit RoboCup style ;)

Reply:
I searched for this but couldn't find anything. There is a library for roslaunch but not for rosnode AFAICT. Yes, this is very RoboCup style;  drastic times call for drastic measures.